### ISSUE TITLE
Feedback homework week 3

### ISSUE THREAD

**Issue by remarcmij** [WaelAlnadeem/hyf-javascript3#4]
Hi Wael, here is my feedback on your week 3 homework (`OOP and ES6 classes` folder).

When I load up your `index.html` file everything seems to be in working order. You app is responsive, although the aspect ratio of the avatars is either stretched out horizontally (on a wide-screen) or squeezed (on a mobile phone).

**1.** Your `sortList()` function looks like a straight copy from Isaac's earlier work. We expect you to deliver your own, original work. If you and Isaac worked together on this function, please mention this in your code. Otherwise people can mistake you for a cheater.

This is the feedback I gave to Isaac on this `sortList` function. With some modifications you can apply it to your own code:

In your `manipulateSelect()` function you create a `<select>` element and add `<option>` elements in the order of the repository data returned by GitHub. Then, you use your `sortList()` function to correct the `text` and `value` properties of each `<option>` element to make them appear in alphabetically sorted order. This seems rather convoluted. Why not sort the repo array before creating `<option>` statements in the first place, like shown here.

```js
function manipulateSelect(repos) {

  repos.sort((a, b) => a.name.localeCompare(b.name)); // <== added

  const select = createAndAppend('select', document.getElementById('header'));

  // ...

    repos.forEach((repo, i) => {
    createAndAppend('option', select, {
      html: repos[i].name,
      value: i
    });
  });

  // sortList(select); // <== removed

  // ...
}
```

So instead of sorting afterwards (and correcting what was the wrong order initially), do it directly on the data as you receive it (without creating something in the wrong order first and then correcting it). You only need a single line for the sort (but of course you need to know how you can use `.sort()` in combination with the `localeCompare()` string method).

You can now completely get rid of your `sortList` function.

**2.** The `initialize()` method of your `View` class does not handle errors from `fetchJSON()`. If I change in the `APIUrl` `orgs` to `orgsx` I get this error message in the console:

```
app.js:135 Uncaught (in promise) Error: Not Found
    at XMLHttpRequest.req.onreadystatechange (app.js:135)
```

You need to add a `try`/`catch` block in `initialize()` and render the error to the web page.

**3.** The following code snippet from the `initialize()` method contains an unnecessary intermediate variable:

```js
let repos;
const data = await this.fetchJSON(APIUrl);
repos = data;
repos.forEach((repo, index) => {
  createAndAppend('option', select, { html: repo.name, value: index });
});
```

You can define and assign `repos` in one line:

```js
const repos = await this.fetchJSON(APIUrl);
repos.forEach((repo, index) => {
  createAndAppend('option', select, { html: repo.name, value: index });
});
```

**4.** In the assignment, it was suggested that you create a method `fetchContributors()` in the `Repository` class. Because the `Repository` class has all the information needed (in your case in `this._infoData`) to manage itself completely, it can fetch the contributors without an external caller needing to tell it what URL to use. So this `fetchContributors()` method could look like this:

```js
fetchContributors() {
  return fetchJSON(this._infoData.contributors_url);
}
```

However, to make this work you would need to move the `fetchJSON()` function out of the `View` class and make it a stand-alone function again, similar to `createAndAppend()`.

With this `fetchContributors()` method in the `Repository` class your `render()` method of the `View` class could look like this:

```js
async render(repoData) {
  try {
    // ...
    const repository = new Repository(repoData);
    const data = await repository.fetchContributors();
    const contributors = new Contributor(data);
    repository.render(details);
    contributors.render(contributor);
  }
  catch (err) {
    root.innerHTML = err.message;
  }
}
```

Now, your `View` class can remain completely unaware about the internal details of the repository information (that's a good thing).

BTW: avoid the use of `alert()`. It is too much "in your face" for most users. Rendering the error in the web page is good enough, although you want to style it as an error (e.g. red text or something like that.)

**5.** There is a problem with your use of `window.onload`. The way you coded it does not actually cause your code to be executed when the page is fully loaded. Instead it is executed immediately when the JavaScript engine loads the script file, which may be too early. If I move the `<script src="app.js">` tag from `<body>` to `<header>` (which is a valid thing to do), you app no longer works. I get this error in the console:

```
Uncaught (in promise) TypeError: Cannot read property 'appendChild' of null
    at createAndAppend (app.js:112)
    at View.mainPage (app.js:54)
    at new View (app.js:47)
    at app.js:149
```

The `window.onload` property must be assigned a function. That function will be called when the page is completely loaded. For instance, you could do this (using an arrow function):

```js
window.onload = () => new View();
```

**6.** Please pay attention on where you use blank lines. Every blank line that you use should be there for a reason. Use a single blank line between function and class definitions. Use a single blank line between blocks of related code where needed to improve readability. Remove all blank lines that do not fall into one of these categories.

Bottom line: I think you have a reasonable grasp of the concepts we covered in JS3. But please don't make teachers doubt about your abilities by copying code from others without an explanation.
